Feature/context sensitive ait merge 3

martin-cs opened the pull request:
The next part of the history sensitivity patch set.  Unlike previous PRs, this really starts changing things.  The key commits split out storage of domains into a separate layer and replace locations with histories (abstract sequences of histories).  Other commits perform supporting refactoring, add examples that use the new functionality, etc.  The changes are quite major however they should not alter the externally visible behaviour of the abstract interpreter at all!

As before, I would recommend reviewing a patch at a time.

Commit:
Add the history interface and the most basic implementation of it.

This is a key step in separating the idea of domains and locations.
It allows progress to be tracked independently of changes to the
domains and gives the basis for more sophisticated handling of
control-flow (such as context sensitivity and loop unwinding).

Commit:
Add a function to compute an edge (i.e. one step) in the AI.

This is part of a refactoring so that calling of transform and merge
aren't duplicated all over the code base.

Commit:
Refactor visit and call_function to use compute_edge.

Based on this refactoring, it might look like the interface of
compute_edge is wrong; later patches will make this clearer.

Commit:
Add a layer above ai_baset that handles domain specific things.

This separates out handling of domain specific operations and their
storage.

Commit:
Refactoring allows on-the-fly creation of domains as needed.

Commit:
Use the correct type names.

Commit:
Add location-insensitive analysis, i.e. one domain per function.

Commit:
Fix a typo in goto-analyzer's option handling.

Almost certainly irrelevant as the option isn't implemented anyway...

Commit:
Specialised output functions for the location-insensitive domain.

Commit:
Add support for lighter-weight constant and interval analysis.

Commit:
Remove const qualifiers for dep_graph as access may modify.

So we can remove the up-front initialisation of domains, we need
to tweak the use cases where operator [] is used to access
domains that may have not been reached.  Removing the const
qualifier is sufficient to have access to the get_state version
of operator [].

Commit:
A basic test of the location insensitive analysis.

smowton commented on src/analyses/ai.h:
```diff
@@ -289,34 +289,88 @@ class ai_baset
+  }
 
-    return it->second;
+  void clear() override
```
? Does this override only to immediately defer to the thing it overrode?

martin-cs replied:
:-\
Fixed.

smowton commented on src/analyses/ai.h:
```diff
@@ -289,34 +289,88 @@ class ai_baset
+  }
+
+private:
+  // to enforce that domainT is derived from ai_domain_baset
```
You can do this better in C++11 -- `static_assert(std::is_base_of<baset, derviedt>::value, "Must derive from ai_domaint")`

martin-cs replied:
Done.

smowton commented on src/analyses/ai.h:
```diff
@@ -332,57 +386,62 @@ class ait:public ai_baset
     const namespacet &ns) override
   {
-    sequential_fixedpoint(goto_functions, ns);
+    this->sequential_fixedpoint(goto_functions, ns);
```
? why add the explicit `this`?

martin-cs replied:
Ummm... seemed like a good idea at the time.  Removed to make a smaller diff.

martin-cs replied:
Ah, no, without it, it gives a warning about ambiguity between the templated parent and normal methods.

smowton commented on src/analyses/ai.h:
```diff
@@ -332,57 +386,62 @@ class ait:public ai_baset
-  {
-    return util_make_unique<domainT>(static_cast<const domainT &>(s));
-  }
+/// Connects up the methods for sequential analysis
```
I don't really understand anything by this description, elaborate a bit?

martin-cs replied:
Reworded to "Use the sequential analysis fixpoint to perform the analysis."  It is just getting the general fixpoint calling the sequential one.  TBH the sequential / concurrent part could be improved in general (for example auto-detecting threaded programs and doing the right thing).

smowton commented on src/analyses/ai.h:
```diff
@@ -332,57 +386,62 @@ class ait:public ai_baset
+/// Connects up the methods for sequential analysis
+/// aiT is expected to be derived from ai_baset
+template <typename aiT>
+class sequential_analysist : public aiT
```
Name seems likely to clash, is this is a `sequential_ai_xyzt` perhaps?

martin-cs replied:
Well... it's the thing that adds sequential to any domain in the same way that concurrency_aware_ait adds concurrency.  I could just omit this completely and make sequential the default.

martin-cs replied:
Thinking about it; it makes more sense to remove this class and fold into ai_storaget.

smowton commented on src/analyses/ai.h:
```diff
@@ -423,6 +491,41 @@ class concurrency_aware_ait:public ait<domainT>
+/// Specific kinds of analyzer
+/// Also examples of how to combine analysis type, history and domain.
+
+/// ait : sequential, location sensitive
```
Perhaps elaborate on `sequential`? At first sight this looks like a strategy for visiting locations rather than the antonym of "concurrent"

martin-cs replied:
These two classes are to maintain the existing interface.  I've tweaked the wording slightly.

smowton commented on src/analyses/ai.cpp:
```diff
@@ -313,6 +313,33 @@ bool ai_baset::visit(
   return new_data;
 }
 
+bool ai_baset::compute_edge(
```
`visit_edge` maybe?

martin-cs replied:
It is so.

smowton commented on src/analyses/ai.cpp:
```diff
@@ -264,49 +264,45 @@ bool ai_baset::visit(
-    if(l->is_function_call() &&
-       !goto_functions.function_map.empty())
+    DATA_INVARIANT(
+      goto_program.get_successors(l).size() == 1,
```
Can probably get this in one with something like `goto_program.get_successors(l) == { std::next(l) }`

martin-cs replied:
I have a feeling that this gets modified by a later patch.  This would explain some things being a bit more broken out than expected.

smowton commented on src/analyses/ai.cpp:
```diff
@@ -264,49 +264,45 @@ bool ai_baset::visit(
     {
-      new_data=true;
-      put_in_working_set(working_set, to_l);
+      if(to_l == goto_program.instructions.end())
```
I think this might imply `l->is_end_function()`?

martin-cs replied:
IIRC END_FUNCTION has no successors.  This was a branch from the previous implementation and I'm keeping it as I want to change the external behaviour and results of the analyser as little as possible.

smowton commented on src/analyses/ai.cpp:
```diff
@@ -402,11 +387,9 @@ bool ai_baset::do_function_call(
 
-    std::unique_ptr<statet> tmp_state(make_temporary_state(end_state));
-    tmp_state->transform(l_end, l_return, *this, ns);
+    working_sett working_set; // Redundant; visit will add l_return
```
How about passing a pointer then (or optional<ref_wrapper<...>> if we really must) to dodge the temporaries

martin-cs replied:
These are only added to get the code to compile before the de-recursion patch; they'll be gone in a day or two.

smowton commented on src/goto-instrument/full_slicer.cpp:
```diff
@@ -20,9 +20,11 @@ Author: Daniel Kroening, kroening@kroening.com
   const dep_node_to_cfgt &dep_node_to_cfg)
 {
+  // node.PC may not have been previously explored
+  // so dep_graph cannot be const as this may create new domains
```
This might be a sensible use of `mutable`, if the class presents the illusion of storing `bottom` for all domains, but in fact populates the domain map on demand, it nontheless appears to behave like a const operation from the outside.

martin-cs replied:
Yes at the moment.  If your patch for storing only merge points gets the obvious extension to do on-the-fly computation of missing domains then it will start to be a serious amount of computation and I'd feel bad about hiding that behind a mutable.

peterschrammel replied:
new states

smowton commented on src/analyses/ai_history.h:
```diff
@@ -0,0 +1,215 @@
+  bool operator<(const history_ptrt &op) const;
+  bool operator==(const history_ptrt &op) const;
+
+  const ai_history_baset &operator*(void)const
```
nitpick: space before `const`

martin-cs replied:
Fixed.

martin-cs replied:
Interestingly, this is something that clang-format introduces!

smowton commented on src/analyses/ai_history.h:
```diff
@@ -0,0 +1,215 @@
+/// explores the program and for context sensitive analyses it is
+/// also used as the key for the state map.
+
+/// We make heavy use of reference counted pointers to history objects
```
I think this will attach to `class ai_history_baset` whereas I guess you want this to be file-level doc

martin-cs replied:
I actually meant it to be connect to ```class ai_history_baset``` but I think file level would be better.  Refactored.

peterschrammel replied:
>  Refactored.

Where?

smowton commented on src/analyses/ai_history.h:
```diff
@@ -0,0 +1,215 @@
+  {
+  }
+
+  enum step_status
```
Definitely document the meanings of these states here, or refer me to the function below that elaborates their meaning

smowton replied:
Also I guess `step_statust`, and should probably use `enum class`

martin-cs replied:
Fixed.

martin-cs replied:
I've expanded the documentation below.

smowton commented on src/analyses/ai_history.h:
```diff
@@ -0,0 +1,215 @@
+
+  virtual step_returnt step(locationt to, history_sett &others) const = 0;
+
+  /// PRECONDITION(to.id_dereferenceable());
```
typo `is_deref`

martin-cs replied:
Fixed.

smowton commented on src/analyses/ai_history.h:
```diff
@@ -0,0 +1,215 @@
+  ///     POSTCONDITION(IMPLIES(result.first == MERGED,
+  ///                           result.second is an element of others and
+  ///                           others is unchanged));
+  ///  3. Block this flow of execution (this may omit some program traces)
```
Elaborate a bit on "block"?

martin-cs replied:
It causes the abstract interpreter to skip that edge.  Documented.

smowton commented on src/analyses/ai_history.h:
```diff
@@ -0,0 +1,215 @@
+  /// POSTCONDITION(return value is dereferenceable)
+  virtual const locationt &current_location(void) const = 0;
+
+  /// Domains with a substantial height may need to widen when merging
```
An example might help here?

martin-cs replied:
Intervals!  It's to avoid the ```for (i = 0; i < BIG_NUMBER; ++i) {}``` problem we have at the moment.  I've mentioned intervals in the comment.

smowton commented on src/analyses/ai_history.h:
```diff
@@ -0,0 +1,215 @@
+    return false;
+  }
+
+  virtual void output(std::ostream &out) const
```
Maybe pure virtual?

martin-cs replied:
Probably better.

smowton commented on src/analyses/ai_history.h:
```diff
@@ -0,0 +1,215 @@
+/// The common case of history is to only care about where you are now,
+/// not how you got there!
+/// Invariants are not checkable due to C++...
+class ahistoricalt : public ai_history_baset
```
`ai_history_insensitivet`?

smowton commented on src/analyses/ai_history.h:
```diff
@@ -0,0 +1,215 @@
+  {
+  }
+
+  step_returnt step(locationt to, history_sett &others) const override
```
Document intended strategy

martin-cs replied:
Done.

smowton commented on src/analyses/ai_history.h:
```diff
@@ -0,0 +1,215 @@
+  }
+};
+
+#undef IMPLIES
```
looks like IMPLIES is only used in comments, so no need to define?

martin-cs replied:
Ooops.  Fixed.

smowton commented on src/analyses/ai.cpp:
```diff
@@ -216,16 +217,18 @@ void ai_baset::finalize()
-  working_sett::iterator i=working_set.begin();
-  locationt l=i->second;
-  working_set.erase(i);
+  // STL guarantees this will be key minimal
```
Just "working_set is sorted by location number"?

martin-cs replied:
It isn't necessarily.  It is sorted by whatever ordering the history uses.  This will matter in later patches.

smowton commented on src/analyses/ai.cpp:
```diff
@@ -381,20 +405,27 @@ bool ai_baset::do_function_call(
+    if(end_history.size() == 0)
+      return false; // Function does not return
+
+    DATA_INVARIANT(end_history.size() == 1, "At most one history per location");
```
Specifically for the end-function state? (Rephrase as a "should" statement, so the invariant dump message is clear)

martin-cs replied:
This is a temporary patch; it's only true for ahistorical and it will get removed in a few days.

smowton commented on src/analyses/ai.h:
```diff
@@ -218,18 +221,16 @@ class ai_baset
     const goto_programt &goto_program,
     const irep_idt &identifier) const;
 
+  // the work-queue is sorted using the history's less operator
```
Aha, in which case it would be highly advisable for history's operator< to order primarily by location number for efficiency?

martin-cs replied:
Yes when the histories are in the same function.  Originally I implemented the histories and the de-recursion in the same patch but no-one would ever want to review that so I separated them.  Some details like this are important but are subject to rapid change.

smowton commented on src/analyses/ai.h:
```diff
@@ -254,123 +255,168 @@ class ai_baset
+  typedef std::map<locationt, working_sett> history_mapt;
+  history_mapt history_map;
+
+  history_ptrt start_history(locationt bang) override
```
...bang?

martin-cs replied:
History starts with the big bang.

smowton commented on src/goto-analyzer/goto_analyzer_parse_options.cpp:
```diff
@@ -219,7 +219,7 @@ void goto_analyzer_parse_optionst::get_command_line_options(optionst &options)
     options.set_option("domain set", true);
   }
-  else if(cmdline.isset("(show-non-null)"))
+  else if(cmdline.isset("show-non-null"))
```
Anything to do with the PR? If not submit this on its own

martin-cs replied:
It's not but it's a 1-line change on a broken feature which I noticed while editing nearby code so I thought it would be just less hassle to fit it in here.

smowton commented on src/goto-analyzer/goto_analyzer_parse_options.cpp:
```diff
@@ -810,8 +825,9 @@ void goto_analyzer_parse_optionst::help()
     // NOLINTNEXTLINE(whitespace/line_length)
-    " --location-sensitive         use location-sensitive abstract interpreter\n"
-    " --concurrent                 use concurrency-aware abstract interpreter\n"
+    " --location-sensitive         one domain per location\n"
```
What happens if neither location-sen nor location-insen is supplied?

martin-cs replied:
It defaults to location-sensitive ( https://github.com/diffblue/cbmc/blob/develop/src/goto-analyzer/goto_analyzer_parse_options.cpp#L246 ) as the "expected" style of abstract interpretation.

smowton replied:
Mark the default in the help text in that case

peterschrammel replied:
domain -> abstract state,  please

peterschrammel replied:
Maybe put options and help text into defines so that they can be reused in janalyzer

smowton commented on src/goto-analyzer/goto_analyzer_parse_options.cpp:
```diff
@@ -810,8 +825,9 @@ void goto_analyzer_parse_optionst::help()
-    " --concurrent                 use concurrency-aware abstract interpreter\n"
+    " --location-sensitive         one domain per location\n"
+    " --location-insensitive       one domain per function\n"
+    " --concurrent                 one domain per location and concurrency analysis\n"
```
Is local-insensitive / concurrency-aware a  sensible combo?

martin-cs replied:
Yes but at the moment none of the domains that work with ```--concurrent``` are integrated into goto-analyzer so it's more of a TODO than a pressing issue.

martin-cs commented:
Many thanks to @smowton for his rapid and insightful review.  I think I've addressed all the comments and have pushed a version of the PR with updates and fixes.

martin-cs commented:
Broken because I am a fool; will fix ASAP.

smowton commented on src/analyses/ai.cpp:
```diff
@@ -239,126 +240,159 @@ bool ai_baset::fixedpoint(
     f_it->second;
 
+  locationt l_return = std::next(l_call);
+  // DATA_INVARIANT(l_return.is_dereferenceable(),
```
Could make this real like the != end() check below?

martin-cs replied:
I don't think so because you don't have the caller goto_program object and so have nothing to ```.end()```.  (Another instance of the "maybe we shouldn't use iterators to describe locations" ?)

smowton replied:
Ah apologies, I didn't realise that was the callee

smowton commented on src/analyses/ai.cpp:
```diff
@@ -239,126 +240,159 @@ bool ai_baset::fixedpoint(
+    working_sett working_set; // Redundant; fixpoint will add l_begin
+    INVARIANT(
+      l_begin != goto_function.body.instructions.end(),
+      "Have checked body_available()");
```
Rephrase as a should statement

martin-cs replied:
Done.

smowton commented on src/analyses/ai_history.h:
```diff
@@ -0,0 +1,229 @@
+  /// Step may do one of four things :
+  ///  1. Create a new history object in others and return a pointer to it.
+  ///     Always adds the history to the worklist.
+  ///     POSTCONDITION(IMPLIES(result.first ==step_statust::NEW_FORCE_CONTINUE,
```
== spacing

smowton replied:
Good place to say why you'd do such a thing?

martin-cs replied:
Eagle eyes for spotting the trick I used to get this in 80 columns.  It's ugly but splitting it over two lines is worse IMHO.

martin-cs replied:
Reasons... it's a bit complicated.  This bit is still evolving a little as I'm working on other contexts.  One use case is the one given in ahistorical, where you don't want to not explore parts of the program because you're not enlarging the domains (for example, location insensitive or the fully flow insensitive).  You probably want to do this if you are creating the first history in a set.  If your history has other external effects (like running the binary in sync with the AI) then again you want to force continue.

peterschrammel replied:
Please check doxygen whether these comments appear somewhere sensible.

smowton commented on src/analyses/ai_history.h:
```diff
@@ -0,0 +1,229 @@
+
+      const auto it = others.begin();
+      INVARIANT(
+        *(it) == next,
```
We're allocating a new `ahistoricalt` just to check this invariant? Maybe only `new` on the others.empty() path and allocate on the stack here?

martin-cs replied:
Fixed.

smowton reviewed (commented):
A few more comments

Commit:
Add the default options to the help text output.

Commit:
Replace locationt's with historys in the abstract interpreter.

This should not alter the behaviour of any of the analyzers.
However it enables domains that track less than one domain per
location and is the foundation for having more than one domain
per location.

martin-cs commented:
All issues addressed and new version (one that might even compile!) pushed.

peterschrammel commented on src/analyses/ai.cpp:
```diff
@@ -313,6 +313,33 @@ bool ai_baset::visit(
+  const locationt &to_l,
+  const namespacet &ns)
+{
+  // Abstract domains are mutable so we must copy before we transform
```
I don't like the use of the word 'domain' here. Call it abstract _state_.

peterschrammel replied:
Which of the lines below does the copying?

martin-cs replied:
s/domain/state/ done.  The line just below with make_temporary_state is the one that does the copying.

peterschrammel commented on src/analyses/ai.cpp:
```diff
@@ -264,49 +264,45 @@ bool ai_baset::visit(
+      "function calls only have one successor...");
+    DATA_INVARIANT(
+      *(goto_program.get_successors(l).begin()) == std::next(l),
+      "... and it is the next instruction / return location");
```
If this breaks I won't see the first part of the sentence. Please make the error messages independent.

martin-cs replied:
Fair point.  Fixed.

peterschrammel commented on src/analyses/ai.cpp:
```diff
@@ -313,6 +313,33 @@ bool ai_baset::visit(
+bool ai_baset::visit_edge(
+  locationt l,
+  working_sett &working_set,
+  const locationt &to_l,
```
Why const &  here and not above?

peterschrammel commented on src/analyses/ai.h:
```diff
@@ -289,34 +289,98 @@ class ai_baset
+    locationt to,
+    const namespacet &ns) override
+  {
+    throw "not implemented";
```
UNIMPLEMENTED?

martin-cs replied:
Definitely better!

peterschrammel commented on src/analyses/ai.h:
```diff
@@ -289,34 +289,98 @@ class ai_baset
+/// There are several different options of what kind of storage is used for
+/// the domains.
+
+/// location_sensitive_ait stores one domain per location
```
state

martin-cs replied:
Fixed.

peterschrammel commented on src/analyses/ai.h:
```diff
@@ -332,78 +396,53 @@ class ait:public ai_baset
-  {
-    throw "not implemented";
-  }
+  // FIXME : should be private
```
Fix it now?

martin-cs replied:
I should but ... the problem is that some of the older domains access the state map directly.  Particularly the dependency graph domain.  Last time I looked at it, the patch to fix it touched a lot of places and I didn't want this patch set to break various other people's work.  If you say fix it, I will, otherwise I might wait a little bit.

peterschrammel replied:
Ok.

peterschrammel commented on src/analyses/ai.cpp:
```diff
@@ -201,10 +201,8 @@ void ai_baset::initialize(const goto_functionst::goto_functiont &goto_function)
-  forall_goto_program_instructions(i_it, goto_program)
-    get_state(i_it).make_bottom();
+  // Domains are created and set to bottom on access.
+  // So we do not need to set them to be bottom before hand.
```
to bottom beforehand

martin-cs replied:
Fixed butter-fingered foolery.

peterschrammel commented on src/analyses/ai.h:
```diff
@@ -413,6 +413,105 @@ class ai_storaget : public ai_baset
 /// There are several different options of what kind of storage is used for
 /// the domains and how historys map to domains.
 
+/// location_insensitive_ait stores one domain per function
```
state

peterschrammel commented on src/analyses/ai.h:
```diff
@@ -413,6 +413,105 @@ class ai_storaget : public ai_baset
+    return static_cast<domainT &>(get_state(l->function));
+  }
+
+  const domainT &operator[](const locationt &l) const
```
why const & ?

peterschrammel commented on src/analyses/ai.h:
```diff
@@ -413,6 +413,105 @@ class ai_storaget : public ai_baset
+
+protected:
+  // this one creates states, if need be
+  virtual statet &get_state(const irep_idt f)
```
const &

martin-cs replied:
Done.

peterschrammel commented on src/analyses/ai.h:
```diff
@@ -413,6 +413,105 @@ class ai_storaget : public ai_baset
+  }
+
+  // this one just finds states and can be used with a const ai_storage
+  virtual const statet &find_state(const irep_idt f) const
```
const &

martin-cs replied:
Done.

peterschrammel commented on src/analyses/ai.h:
```diff
@@ -413,6 +413,105 @@ class ai_storaget : public ai_baset
+  {
+    typename state_mapt::const_iterator it = state_map.find(f);
+    if(it==state_map.end())
+      throw "failed to find state";
```
DATA_INVARIANT?

martin-cs replied:
PRECONDITION; it is the caller's problem to make sure that get_state was used first.

peterschrammel commented on src/analyses/ai.cpp:
```diff
@@ -237,33 +240,37 @@ bool ai_baset::fixedpoint(
+  {
+    locationt first = goto_program.instructions.begin();
+    put_in_working_set(working_set, start_history(first));
+    /// TODO : should only start history once?
```
Fix me

martin-cs replied:
Fixed in the next patch set; to get one start and function calls you have to de-recursion the AI.

peterschrammel commented on src/analyses/ai.cpp:
```diff
@@ -381,20 +405,27 @@ bool ai_baset::do_function_call(
     assert(l_end->is_end_function());
 
+    working_sett end_history = get_history(l_end);
+    if(end_history.size() == 0)
```
.empty() ?

martin-cs replied:
Fixed.

peterschrammel commented on src/analyses/ai.h:
```diff
@@ -254,92 +255,133 @@ class ai_baset
+    ai_baset::clear();
+  }
+
+  working_sett get_history(locationt &l) override
```
Should the return value be copied, const, or writeable?

martin-cs replied:
Point.  It was a copy due to C++ type system weirdness.  Post refactor, it can be const and now it is!

peterschrammel commented on src/analyses/ai.cpp:
```diff
@@ -190,8 +190,9 @@ void ai_baset::entry_state(const goto_functionst &goto_functions)
 {
-  // The first instruction of 'goto_program' is the entry point
-  get_state(goto_program.instructions.begin()).make_entry();
+  // The first instruction of 'goto_program' is the entry point with no history
```
(commit message) historys -> histories

martin-cs replied:
Will fix.

peterschrammel commented on src/analyses/ai_history.h:
```diff
@@ -0,0 +1,227 @@
+  {
+  }
+
+  enum class step_statust
```
A bit of doxygen explaining the below would be helpful.

martin-cs replied:
Explained in step below.  Is there a way to view the CI generated doxygen to check it has come out OK?

peterschrammel replied:
You can just run `doxygen` in `src` locally and then open `doc/html/index.html`

martin-cs replied:
Done and now it is correctly linked to the right functions.

peterschrammel commented:
Hidden comments are relevant.

martin-cs commented:
Update:

1. One of the reasons the tests are failing is that concurrent fix-point uses the rest of the analyzer in a very different way and has an interesting take on some of the system invariants.  This leads to formally undefined behaviour via iterator manipulation inside various standard data types.  For reasons as yet unknown this crashes the debug mode rather than giving an error.

2. Two fix-up things still to do for @peterschrammel : 

- turn help messages and options into macros
- be consistent about the use of locationt

Then (and with a rebase et al.) we should be good to go.

martin-cs commented:
Rewritten into later patches.
